### PR TITLE
priv_debugGetStateRoot bug refactorings.

### LINKED ISSUE #1596: PrivDebugGetStateRoot doesn't work
In PrivDebugGetStateRoot, we receive the privacy group id and a block hash/number to retrieve the state root hash of a privacy group at a specific point in the chain. I’ve stumbled upon this:
    final PrivacyGroup[] privacyGroups =
        privacyController.findPrivacyGroup(
            Collections.singletonList(privacyGroupId),
            enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser())); 
However, the api for finding the privacy group is the following:  privacyController.findPrivacyGroup(List<String> addresses, String enclavePublicKey)
. It feels like we are using the privacyGroupId instead of the participants keys.
Am I missing something or have we got a bug in this method?

Maybe an internal rename to help us in future:
findPrivacyGroup -> findPrivacyGroupByMembers
retrievePrivacyGroup -> findPrivacyGroupById

### PR DESCRIPTION
**Pull request by mark-terry**
Signed-off-by: Mark Terry <mark.terry@consensys.net>

## PR description
Fix for bug in `priv_debugGetStateRoot` method.

## Fixed Issue(s)

Fixes #1596 

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDebugGetStateRoot.java [hunk 1]
```diff
@@ -68,13 +69,19 @@ protected Object resultByBlockNumber(
         enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser());
     LOG.trace("Executing {}", this::getName);
 
-    final PrivacyGroup[] privacyGroups =
-        privacyController.findPrivacyGroup(
-            Collections.singletonList(privacyGroupId),
-            enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser()));
+    final Optional<PrivacyGroup> privacyGroup;
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDebugGetStateRoot.java:72]: To provide better feedback about the RPC parameters. Here we should catch: incorrect privacyGroupId, multi-tenancy validation problems, other invalid/incorrect parameter issues.

The error branch when the response returns will only be triggered if there's a problem retrieving the stateRoot, which, at that point, shouldn't be related to the RPC params... That return code could probably be improved.

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivDebugGetStateRootOnchainGroupAcceptanceTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+    waitForBlockHeight(aliceNode, 10);
+    waitForBlockHeight(bobNode, 10);
+
+    final Hash aliceResult1 =
+        aliceNode.execute(privacyTransactions.debugGetStateRoot(privacyGroupId, "1")).getResult();
+    final Hash bobResultInt1 =
+        bobNode.execute(privacyTransactions.debugGetStateRoot(privacyGroupId, "1")).getResult();
+
+    assertThat(aliceResult1).isEqualTo(bobResultInt1);
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivDebugGetStateRootOnchainGroupAcceptanceTest.java:115]: A consistent Hash is returned until some block after the group has been created. Seems, if the requested block is before group creation, it looks back to the earliest state root.

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivDebugGetStateRootOnchainGroupAcceptanceTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    final Hash bobStateRootId =
+        bobNode
+            .execute(privacyTransactions.debugGetStateRoot(privacyGroupId, "latest"))
+            .getResult();
+
+    assertThat(aliceStateRootId).isEqualTo(bobStateRootId);
+  }
+
+  @Test
+  public void nodesNotInGroupShouldReturnError() {
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivDebugGetStateRootOnchainGroupAcceptanceTest.java:68]: I think you're right. Will remove.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java [hunk 4]
```diff
@@ -74,17 +74,21 @@ long determineEeaNonce(
   Optional<String> buildAndSendAddPayload(
       PrivateTransaction privateTransaction, Bytes32 privacyGroupId, String enclaveKey);
 
-  Optional<PrivacyGroup> retrieveOffChainPrivacyGroup(String toBase64String, String enclaveKey);
+  Optional<PrivacyGroup> findOffChainPrivacyGroupByGroupId(
+      String toBase64String, String enclaveKey);
 
-  List<PrivacyGroup> findOnChainPrivacyGroup(List<String> asList, String enclaveKey);
+  Optional<PrivacyGroup> findPrivacyGroupByGroupId(
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:80]: should this be findOnChainPrivacyGroupByGroupId?

**Review by macfarla — COMMENTED** [review]
couple of spots where naming could be more consistent

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java [hunk 5]
```diff
@@ -259,13 +260,25 @@ public long determineBesuNonce(
   }
 
   @Override
-  public Optional<PrivacyGroup> retrieveOffChainPrivacyGroup(
+  public Optional<PrivacyGroup> findPrivacyGroupByGroupId(
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java:263]: This method is checking both on and off chain groups. It first checks the offchain, then tries onchain if the enclave indicates the offchain group doesn't exist (via an exception.)

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java (hunk 4 above)
Comment (mark-terry) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:80]: See reply from earlier. The `findOnChainPrivacyGroupByGroupId` is in the implementation, not the interface.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (macfarla) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java:263]: got it

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java (hunk 4 above)
Comment (macfarla) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:80]: 👍

**Review by lucassaldanha — APPROVED** [review]
LGTM

**Pull request merged**